### ISSUE TITLE
can't set a chanel encryption key

### ISSUE THREAD

**Issue by modosc** [ably/ably-js-react-native#11]
we are successfully using encryption keys for a ruby backend and js frontend. the same code doesn't work in react-native though, and it seemingly stems from the `buffer` polyfill.

on the web js side this code works:
```javascript
      channel = ably.channels.get(channelName, { cipher: { key: Buffer.from(ABLY_ENCRYPTION_KEY, 'hex') } })
```

in react-native this doesn't work because there's no `buffer` available. if i use the [buffer polyfill](https://github.com/feross/buffer) and the above code i get this error from ably:
```
Unhandled promise rejection TypeError: Cannot read property 'randomBytes' of undefined
    at randomBytes (index.js:14)
    at Object.getRandomValues (ably-reactnative.js:3128)
    at generateRandom (ably-reactnative.js:3166)
    at CBCCipher.encrypt (ably-reactnative.js:3396)
    at Function.Message.encrypt (ably-reactnative.js:5194)
    at Function.Message.encode (ably-reactnative.js:5219)
    at RealtimePresence._enterOrUpdateClient (ably-reactnative.js:11153)
    at RealtimePresence.enter (ably-reactnative.js:11104)
```
if i _don't_ use the `buffer` polyfill i get this message instead:
```
Unhandled promise rejection TypeError: function Buffer(arg) {
    if (!(this instanceof Buffer)) {
      if (arguments.length > 1) return new Buffer(a...<omitted>... } is not a function
    at Function.from (native)
    at blob:file:///38dfb449-803f-4eda-97b2-f4b288c80e8f:169209:25
    at run (blob:file:///38dfb449-803f-4eda-97b2-f4b288c80e8f:151305:24)
    at blob:file:///38dfb449-803f-4eda-97b2-f4b288c80e8f:151326:9
    at flush (blob:file:///38dfb449-803f-4eda-97b2-f4b288c80e8f:151675:11)
    at tryCallOne (blob:file:///38dfb449-803f-4eda-97b2-f4b288c80e8f:3049:14)
    at blob:file:///38dfb449-803f-4eda-97b2-f4b288c80e8f:3150:17
    at blob:file:///38dfb449-803f-4eda-97b2-f4b288c80e8f:24267:21
    at _callTimer (blob:file:///38dfb449-803f-4eda-97b2-f4b288c80e8f:24156:9)
    at _callImmediatesPass (blob:file:///38dfb449-803f-4eda-97b2-f4b288c80e8f:24192:9)
reactConsoleErrorHandler @ ExceptionsManager.js:84
console.error @ YellowBox.js:59
(anonymous) @ es6.promise.js:110
module.exports @ _perform.js:3
(anonymous) @ es6.promise.js:104
(anonymous) @ JSTimers.js:295
_callTimer @ JSTimers.js:152
_callImmediatesPass @ JSTimers.js:200
callImmediates @ JSTimers.js:464
__callImmediates @ MessageQueue.js:320
(anonymous) @ MessageQueue.js:135
__guard @ MessageQueue.js:297
flushedQueue @ MessageQueue.js:134
callFunctionReturnFlushedQueue @ MessageQueue.js:109
e @ RNDebuggerWorker.js:1
12:56:27.965 
```

the docs say that base64 encoded strings should be supported so i tried this but this also fails by incorrectly identifying the length of the key and erroring out (keep in mind, this is the same key that we're using without issue in ruby/browser js):
```
Unhandled promise rejection Error: Unsupported key length 384 for aes-cbc encryption. Encryption key must be 128 or 256 bits (16 or 32 ASCII characters)
```
i've tried encoding the string with multiple base64 encoders. 

any clues here?

**Comment by modosc** [comment]
i think the problem is the `randombytes` package - i realized the `buffer` polyfill is working correctly above because i finally managed to pass a base64 encoded string that made the checks happy. after that happens i get the same error:
```
Unhandled promise rejection TypeError: Cannot read property 'randomBytes' of undefined
    at randomBytes (index.js:14)
    at Object.getRandomValues (ably-reactnative.js:3128)
    at generateRandom (ably-reactnative.js:3166)
    at CBCCipher.encrypt (ably-reactnative.js:3396)
    at Function.Message.encrypt (ably-reactnative.js:5194)
    at Function.Message.encode (ably-reactnative.js:5219)
    at RealtimePresence._enterOrUpdateClient (ably-reactnative.js:11153)
    at RealtimePresence.enterClient (ably-reactnative.js:11115)
```
is there something extra i need to do with the version of `react-native-randombytes` that this pulls in? `react-native link` says it's linked, i'm not sure what else to check?

**Comment by SimonWoolf** [comment]
You don't need a Buffer polyfill (at least for Ably) -- the ably-react-native library doesn't use `Buffer`s or expect them as arguments. You can pass binary in as an ArrayBuffer (I think you can just call `buf.buffer` on a Buffer to get a copy that's exposed as an ArrayBuffer), or use a base64-encoded string (as you've now got working).

Re `Cannot read property 'randomBytes' of undefined`, that's odd. The line it's referring to is presumably `require('react-native-randombytes').randomBytes` (which was at line 3128 at around v1.0.13), so `require('react-native-randombytes')` is just returning undefined. So for some reason it isn't finding the dependency. You could verify this by just doing that require in your own code and checking whether it works. Assuming not, that'd be some issue with how deps are being pulled in in your app. Check the `npm install` log to see if it has any errors, I guess. (That dep is slightly different from other deps in that it's fetched from github -- is it possible your network is somehow block the git protocol when it tries to fetch the dependency, maybe?)

**Comment by modosc** [comment]
@SimonWoolf thanks for your response. i double checked and it looks like `react-native link` isn't picking up the `react-native-randombytes` dependency. i installed your fork manually and it does get picked up and linked but the build fails:
```
/Users/jon/Library/Developer/Xcode/DerivedData/swoop-gqarxdnqafqgyhfuqjkurceehjee/Build/Products/Debug-iphonesimulator/include/React/RCTBridgeModule.h:52:16: Redefinition of 'RCTMethodInfo'
/Users/jon/git/mobile/node_modules/react-native-randombytes/RNRandomBytes.m:9:9: In file included from /Users/jon/git/mobile/node_modules/react-native-randombytes/RNRandomBytes.m:9:
/Users/jon/git/mobile/node_modules/react-native-randombytes/RNRandomBytes.h:11:9: In file included from /Users/jon/git/mobile/node_modules/react-native-randombytes/RNRandomBytes.h:11:
/react-native/React/Base/RCTBridge.h:11:9: In file included from ../react-native/React/Base/RCTBridge.h:11:
/react-native/React/Base/RCTBridgeModule.h:52:16: Previous definition is here
/Users/jon/Library/Developer/Xcode/DerivedData/swoop-gqarxdnqafqgyhfuqjkurceehjee/Build/Products/Debug-iphonesimulator/include/React/RCTBridgeModule.h:56:3: Typedef redefinition with different types ('struct (anonymous struct at /Users/jon/Library/Developer/Xcode/DerivedData/swoop-gqarxdnqafqgyhfuqjkurceehjee/Build/Products/Debug-iphonesimulator/include/React/RCTBridgeModule.h:52:16)' vs 'struct RCTMethodInfo')
/Users/jon/git/mobile/node_modules/react-native-randombytes/RNRandomBytes.m:9:9: In file included from /Users/jon/git/mobile/node_modules/react-native-randombytes/RNRandomBytes.m:9:
/Users/jon/git/mobile/node_modules/react-native-randombytes/RNRandomBytes.h:11:9: In file included from /Users/jon/git/mobile/node_modules/react-native-randombytes/RNRandomBytes.h:11:
/react-native/React/Base/RCTBridge.h:11:9: In file included from ../react-native/React/Base/RCTBridge.h:11:
/react-native/React/Base/RCTBridgeModule.h:56:3: Previous definition is here
```
progress i guess? we are using `react-native-0.57.8`

**Comment by modosc** [comment]
ah, i think it's this:

https://github.com/mvayngrib/react-native-randombytes/pull/28

i opened a pr with that change - using this things build and encryption works.

**Comment by SimonWoolf** [comment]
@modosc awesome, thanks for looking into it.
Though since it looks like the issue is with requiring sjcl rather than just installing it, I'm wondering if a better solution would be to switch back to the upstream react-native-randombytes and, in the library, call `require('react-native').NativeModules.RNRandomBytes`, per https://github.com/mvayngrib/react-native-randombytes/issues/20#issue-309718048 . If I make a test release that does that, could you try it out and let me know if it works for you?

If it doesn't, I'll merge your PR and do a release with that.

**Comment by SimonWoolf** [comment]
@modosc Could you try out `ably-react-native@1.1.1-beta.0` (under the `beta` tag) and see if that works for you? That switches to upstream react-native-randombytes, with a few changes to ably-js. If that works well for you, I'll do a proper release with it.

Thanks :slightly_smiling_face:

**Comment by modosc** [comment]
@SimonWoolf it works!  i still have to manually do `react-native link react-native-randombytes` (even though it's linking up your version) - if i don't then the same error occurs.

btw i'm using `yarn`, not `npm`, wonder if that's got something to do with it? also we're currently on `node 10.15` btw but the same issues happened on `8.12` also.

**Comment by SimonWoolf** [comment]
Released v1.1.1 with the above fix (pulling in ably-js v.1.1.3).

I'm not sure what's up with your having to manually `link` the random-bytes dependency, but I'm not a react-native developer. I guess in theory I could run `react-native link react-native-randombytes` in a `postinstall` script, but that seems like the wrong thing to do -- the package shouldn't be making assumptions about either the presence of the react-native cli or what the relative path is to the main project. Maybe I should just note it in the readme in case anyone else runs into the same issues...

**Comment by modosc** [comment]
things are working now (except the manual linking step) - i think we can close this out?
